contracts/DistributionContract.sol

```diff
@@ -1,75 +1,184 @@
+     * @dev filling beneficiary balance
+     *
+     * @param beneficiary_ , contains who's will get tokens
+     * @param amount_ , how many tokens will get
+     * Requirements:
+     *
+     * - `owner` cannot be the zero address.
+     * - `spender` cannot be the zero address.
+     */
+    function addBeneficiary(address beneficiary_, uint256 amount_)
+        external
+        onlyOwner
+    {
+        require(
+            amount_ != 0 && beneficiary_ != address(0),
+            "beneficiary or amount equal to 0"
+        );
+        balanceOfBeneficiares[beneficiary_] += amount_;
+        emit RewardCollect(beneficiary_, amount_, true);
     }
 
-    function emergencyWithdraw (uint256 _amount) public onlyOwner {
-        token.transferFrom(address(this),contractOwner, _amount);
+    /**
+     * @dev decrease balance of beneficiary
+     *
+     * @param beneficiary_ , contains which balance will reduced
+     * @param amount_ , how many tokens will reduced
+     * Requirements:
+     *
+     * - `owner` cannot be the zero address.
+     * - `spender` cannot be the zero address.
+     */
+    function decreaseReward(address beneficiary_, uint256 amount_)
+        external
+        onlyOwner
+    {
+        require(
+            amount_ != 0 && beneficiary_ != address(0),
+            "beneficiary or amount equal to 0"
+        );
+        balanceOfBeneficiares[beneficiary_] -= amount_;
     }
 
-    function lockRewards (uint256 beneficiaryIndex) public onlyOwner {
-        if(beneficiary[beneficiaryIndex].transferStatus == false) {
-            beneficiary[beneficiaryIndex].transferStatus = true;
-        }
-        else if(beneficiary[beneficiaryIndex].transferStatus == true) {
-            beneficiary[beneficiaryIndex].transferStatus = false;
-        }
+    /**
+     * @dev this function allow owner get back tokens
+     *
+     * @param amount_ , how many tokens will withdraw
+     *
+     */
+    function emergencyWithdraw(uint256 amount_) external onlyOwner {
+        _token.safeTransfer(msg.sender, amount_);
     }
 
-    function claim () public {
-        token.transferFrom(address(this),
-        listOfBeneficiaries[msg.sender]._addressOfBeneficiary,
-        listOfBeneficiaries[msg.sender]._reward);
+    /**
+     * @dev change state of 'lockStatus'
+     *
+     * @param lock_ , contains boolean value
+     *
+     * NOTE : this function affects the ability of the beneficiary
+     * to receive tokens
+     */
+    function lockRewards(bool lock_) external onlyOwner {
+        lockStatus = lock_;
+        emit LockStatus(lockStatus);
     }
 
-
-
-}    
\ No newline at end of file
+    /**
+     * @dev this function allows beneficiary to withdraw
+     * their tokens
+     *
+     * NOTE : send all tokens that are available on the balance
+     *
+     */
+    function claim() external {
+        require(lockStatus, "claim is locked !");
+        require(balanceOfBeneficiares[msg.sender] > 0, "balance must be > 0");
```

Comment:
for gas optimization you can check amount variable, on your flow you two time call storage to get balance